docs/typing-indicators.md

```diff
@@ -0,0 +1,146 @@
+protocol.
+
+Right now typing indicators are only used in "Private Messages"
+views.  We implemented them for streams in a very early version of
+the feature, but users found them distracting.  We may re-visit
+this decision, particularly for private streams or streams with
+fewer invitees.
+
+There are two major roles for users in this system:
+
+* The "writing user" is composing a message.
+* The "reading user" is waiting to receive a message (or possibly
+  ready to shift their attention elsewhere).
+
+Any Zulip user can play either one of these roles, and sometimes
+they can be playing both roles at once.  Having said that, you
+can generally understand the system in terms of a single message
+being composed by the "writing user."
+
+On a high level the typing indicators system works like this:
+
+* The client for the "writing user" sends requests to the server.
+* The server broadcasts events to other users.
+* The clients for "reading users" receive events and conditionally
+  show typing indicators, depending on where the users are narrowed.
+
+## Writing user
+
+When a "writing user" starts to compose a message, the client app
+sends a request to an endpoint called `/json/typing` with an `op`
+of `start` and a list of potential message recipients. The JS
+function that facilitates this is called `send_typing_notification_ajax`.
+
+If the "writing user" is composing a long message, we want to send
+repeated updates to the server, so that downstream clients know that the
+user is still typing.  (Zulip messages tend to be longer than
+messages in other chat/text clients.)
+
+We have a small state machine in `static/js/typing_status.js` that
+makes sure subsequent "start" requests get sent out every ten
+seconds.  (This document is intended to describe the high level
+architecture; the actual time values may be tuned in future releases.)
+
+If the "writing user" goes more than five seconds without any text
+input, then we send a request with an `op` of `stop`.  We also send
+"stop" messages when the user closes the compose box.
+
+A common scenario is that a user is just pausing to think for a few
+seconds, but they still intend to finish the message.  Of course,
+that's hard to distinguish from the scenario of the user got pulled
+away from their desk.  For the former case, where the "writing user"
+completes the message with lots of pauses for thinking, a series of
+"start" and "stop" messages may be sent over time.  One way to
+improve our feature here is to tweak the timeout values, but of
+course it is always a tradeoff depending on how quickly people type,
+how long they pause to think, how frequently they get interrupted, etc.
+
+## Server
+
+The server piece of typing notificiations is currently pretty
+straightforward, since we take advantage of Zulip's event system,
+and the server piece here is basically a single Django view
+function with a small bit of library support to send out events
+to clients.
```

Comment:
Link to the event system doc here.

Reply:
Also emphasize that the server logic is stateless; I think that's an important thing to mention up front, and explain that a stateless server is really good, since it limits the possiblities for where bugs in the system could reside.